ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: MPL-2.0
+
+#![allow(dead_code)]
+
+mod fifo_scheduler;
+
+pub use fifo_scheduler::FifoScheduler;
+use spin::Once;
+
+use super::{processor::switch_to_task, task::Task, TaskStatus};
+use crate::{arch::timer::register_callback, prelude::*};
+
+static SCHEDULER: Once<&'static dyn Scheduler<Task>> = Once::new();
+
+/// Injects a scheduler implementation into framework.
+///
+/// This function can only be called once and must be called during the initialization of kernel.
+pub fn inject_scheduler(scheduler: &'static dyn Scheduler) {
+    SCHEDULER.call_once(|| scheduler);
+
+    register_callback(|| {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            if local_rq.update_current(UpdateFlags::Tick) {
+                local_rq.set_should_preempt(true);
+            }
+        })
+    });
+}
+
+/// Adds task to scheduler.
+///
+/// This function may be called with newly spawned task or waked task.
+pub fn add_task(runnable: Arc<Task>, flags: EnqueueFlags) {
+    if !SCHEDULER.is_completed() {
+        let fifo_scheduler = Box::new(FifoScheduler::<Task>::new());
+        let static_scheduler = Box::leak(fifo_scheduler);
+        inject_scheduler(static_scheduler);
+    }
```

Comment:
The function `add_task` will try to init scheduler, which looks incorrect for me. Maybe the code should be put in a more suitable place.

Reply:
I second this concern.

Reply:
This is a compromise because I use `Once` to ensure the `SCHEDULER` can only be set once; I have to check if `SCHEDULER` is already initialized when we are to add the first task and set it to default if not.

However, I do think it's ugly and logically incorrect. I'll see how to fix this. Maybe just revert the `SCHEDULER` to previous implementation?

Reply:
> Maybe just revert the SCHEDULER to previous implementation?

This is also an option. But previous code allows `inject_scheduler` to be called multiple times, which is not correct. Some code still needs to be added to ensure `inject_scheduler` to be called only once.

Reply:
Ensuring `inject_scheduler` can only be called once is not enough; we still have to decide whether to inject default scheduler, which depends on whether the OSTD users provide their own schedulers. So logically the decision should be made by OSTD after initialization of the user kernels, which seems impossible now.

I think there are two options:
- force OSTD users to inject a scheduler (either default or their own)
Logically we OSTD should not be responsible for providing a default policy, which should be done by OSDK or its ecosystem. We may move all the default policy of OSTD's pluggable componenets (e.g., scheduler, page allocator, ...)  to a new crate outside OSTD and enforce OSTD users to explicitly inject each components (either default or their own) in their kernel implementation.
- give up the "called only once" things
We revert the `SCHEDULER` design and we OSTD take the responsibility of migrating tasks to new schedulers when injecting the second time.

Reply:
I vote for the first option. Migrating tasks between schedulers brings unnecessary complexity. I think the current implementation is acceptable as a temporary solution. Eventually, we will force OSTD users to inject a scheduler. If they don't, OSDK can inject a default scheduler for them. Either way, OSTD will have one and only one injected exactly before any tasks are spawned.